### PR TITLE
Optimize MerkleTree.contains from O(n) to O(1)

### PR DESCRIPTION
**Pull request by leanthebean**
## Summary
In the merkle tree file, we have a `contained` method. Currently, there's a linear scan that goes through _**all**_ of the existing nullifiers when determining if a transaction is valid (e.g. not a double spend). As we get more transactions, this linear scan becomes more and more inefficient (e.g. beefy later blocks would take several seconds). 

This change adds another db store that lets you check in O(1) if an element exists in a merkle tree. 

## Testing Plan
Tested it locally and avg block time is about 20ms per transaction per block 

## Breaking Change

This requires a db upgrade. 

 * [x] Yes
 * [ ] No

### CODE REVIEW

File: ironfish/src/consensus/verifier.ts [hunk 1]
```diff
@@ -167,10 +167,8 @@ export class Verifier {
     tx?: IDatabaseTransaction,
   ): Promise<VerificationResult> {
     return this.chain.db.withTransaction(tx, async (tx) => {
-      const noteSize = await this.chain.notes.size(tx)
```
Comment (dguenther) [inline ironfish/src/consensus/verifier.ts:170]: Ah, looks like this might have been a bug, good spot! If possible it might be nice to pull this out into a separate PR so any issues with it aren't conflated with the LeavesIndex cache.

**Review by dguenther — APPROVED** [review]
From a code review standpoint, looks good -- I think @NullSoldier wants to add a perf test to this before it merges, and it'd be nice to run it on a laptop node for a bit to see if any issues happen to pop up.

**Comment by dguenther** [comment]
It looks like there are some test failures in here as well

File: ironfish/src/consensus/verifier.ts (hunk 1 above)
Comment (NullSoldier) [reply ironfish/src/consensus/verifier.ts:170]: Moved that fix here, https://github.com/iron-fish/ironfish/pull/798

**Comment by NullSoldier** [comment]
Here is the perf test @dguenther https://github.com/iron-fish/ironfish/pull/799/files

File: ironfish/src/merkletree/merkletree.ts [hunk 2]
```diff
@@ -339,6 +353,25 @@ export class MerkleTree<E, H, SE extends JsonSerializable, SH extends JsonSerial
...
+      {
+        index: index,
+        element: value.element,
+        merkleHash: value.merkleHash,
+        parentIndex: value.parentIndex,
+      },
+      tx,
+    )
+
+    await this.leavesIndex.put(this.hasher.merkleHash(value.element), index, tx)
```
Comment (NullSoldier) [inline ironfish/src/merkletree/merkletree.ts:372]:
```suggestion
    await this.leavesIndex.put(value.merkleHash, index, tx)
```

File: ironfish/src/consensus/verifier.ts [hunk 3]
```diff
@@ -296,16 +294,15 @@ export class Verifier {
   async hasValidSpends(block: Block, tx?: IDatabaseTransaction): Promise<VerificationResult> {
```
Comment (NullSoldier) [inline ironfish/src/consensus/verifier.ts:294]: The errornously removed code was intending to check for  `the note being spent really existed in the tree at the time it was spent` but was not. Should this still check for that some how? The spend size is the size of the nullifier tree when the commitment was created, not the note tree right?

**Comment by NullSoldier** [comment]
We still need to clean up the index in the `MerkleTree.truncate` function.

File: ironfish/src/consensus/verifier.ts [hunk 4]
```diff
@@ -327,22 +331,21 @@ export class Verifier {
    * spend's spend root.
    *
    * @param spend the spend to be verified
-   * @param size the size of the nullifiers tree at which the spend must not exist
```
Comment (dguenther) [inline ironfish/src/consensus/verifier.ts:330]: This should probably be added back, or we should kill the whole comment so it's at least somewhat consistent

**Review by dguenther — APPROVED** [review]

File: ironfish/src/consensus/verifier.ts (hunk 4 above)
Comment (NullSoldier) [reply ironfish/src/consensus/verifier.ts:330]: Fixed

**Review by NullSoldier — APPROVED** [review]
